admin/server/reports.go

```diff
@@ -53,10 +55,58 @@ func (s *Server) GetReportMeta(ctx context.Context, req *adminv1.GetReportMetaRe
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 
+	opts, err := recreateReportOptionsFromSpec(req.Spec)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	annotations := parseReportAnnotations(req.Spec.Annotations)
+
+	var externalEmails []string
+	for _, email := range opts.EmailRecipients {
+		_, err := s.admin.DB.FindUserByEmail(ctx, email)
+		if err != nil {
+			if errors.Is(err, database.ErrNotFound) {
+				externalEmails = append(externalEmails, email)
+			} else {
+				return nil, status.Errorf(codes.Internal, "failed to find user by email: %s", err.Error())
+			}
+		}
+	}
+
+	externalUsersUrls := make(map[string]*adminv1.GetReportMetaResponse_Urls, len(externalEmails))
+	if len(externalEmails) > 0 {
+		// issue magic tokens for external emails
+		var ownerID *string
+		if annotations.AdminOwnerUserID != "" {
+			ownerID = &annotations.AdminOwnerUserID
+		}
+		emailTokens, err := s.issueMagicTokensForExternalEmails(ctx, proj.OrganizationID, proj.ID, req.Report, ownerID, opts, externalEmails)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to issue magic auth tokens: %s", err.Error())
+		}
+
+		for email, token := range emailTokens {
+			externalUsersUrls[email] = &adminv1.GetReportMetaResponse_Urls{
+				OpenUrl:   s.admin.URLs.WithCustomDomain(org.CustomDomain).ReportOpenExternal(org.Name, proj.Name, req.Report, token, req.ExecutionTime.AsTime()),
+				ExportUrl: s.admin.URLs.WithCustomDomain(org.CustomDomain).ReportExportExternal(org.Name, proj.Name, req.Report, token),
+			}
+		}
+	} else {
+		// cleanup all report tokens if no external emails if found
+		err = s.cleanUpAllReportTokens(ctx, req.Report)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to clean up report tokens: %s", err.Error())
+		}
+	}
```

Comment:
Since `issueMagicTokensForExternalEmails` already revokes tokens for emails not present in the input, it basically works like `cleanUpAllReportTokens` when called with an empty list. So maybe the `else` case can be removed and the `if` statement removed?